_compdemos/python_logging_demo.md

```diff
@@ -0,0 +1,236 @@
+| ERROR | 40 | 'ERROR' |  
+| CRITICAL | 50 | 'CRITICAL' |  
+
+It is also possible to create your own log levels, a process which is described in the Logging Cookbook.
+
+## Where/When to Implement Logging
+I like to log early and log often, especially during the development phase of a project.  In particular, I find often 
+find myself logging the responses to remote calls so I can always have a reference in my log of what the response looked
+like before I start processing it.  When writing REST APIs I often will log the raw request prior to any processing, so
+that when something goes wrong I can look at what data I started with and investigate from there. I've found logging to 
+be invaluable for debugging stateless applications in AWS, as they pick up log messages emitted to STDOUT from your program so I 
+can log state at various points in the application to better understand what might have gone wrong.  In most of my projects I
+create a function to create or retrieve a pre-configured `Logger` instance early on, and use it constantly. I then use 
+environment variables to set the log level of the service to reduce log message volume once the service is more stable.  
+
+## Logging and AWS
+As mentioned before python logging instances will emit their log records to the AWS CloudWatch log group / stream for 
+most of their services (I've seen it wth EC2, ECS, Batch, and Lambda so far). By leveraging structured logging through 
+ python's custom log formatting, you can start to use CloudWatch Insights to query across log groups for messages 
+pertaining to any of the attributes you're logging. 
+
+Log messages a presented in a pretty nice format on CloudWatch:  
+![cloudwatchlogs](./assets/cwlogex.png)  
+
+While it's not necessary to use in order to produce log messages to CloudWatch, there is a project out there called
+[WatchTower](https://github.com/kislyuk/watchtower) which can really help consolidate your logs into particular user-story
+focused log streams that can track logs from across multiple services if they all use `WatchTower` to point to the same log
+group / stream. I've found `WatchTower` helpful for consolidating Lambda logs as by default a new stream is created for 
+every new lambda instance, which can get really confusing, particularly if you have concurrent executions.  I have had 
+some issues (which may have more to do with how I'm creating the `WatchTower` handlers and my application structure) 
+with some of the functions hitting AWS service limits on API calls like Create/DescribeLogGroups and Create/DescribeLogStreams, 
+which can have have an affect on responsiveness and at worst can crash the Lambda call if the exceptions aren't handled properly. 
+This is particularly the case when the service is under a high load. If someone has another solution for this I'd love to hear about it!
+
+
+Give me a shout at `zromer at fredhutch.org` if you have any questions, comments, or corrections you think should be made.
```

Comment:
Thank you for making this! 
Thoughts: either a separate page on logging patterns/a subsection with some quick examples of logging patterns like
main.py: 
> logger = logging.getLogger() #get's the Base Logger, which everything inherits from
> logger.basicConfig(...) #setting this on the Base Logger is inheritable on every other logger that implicitly inherits from it
important_module.py: 
> logger = logging.getLogger(__name__) #this automagically sets the logger to the name of the module (ie: 'important_module.py')
> #additionally, all BasicConfig is inherited from the logger in main.py at this point

Reply:
+1 thanks for contributing this

Reply:
I could definitely whip up some code examples for functions that configure different handlers / levels / filters for loggers that someone could just drag and drop into their projects.